optics-core/src/Optics/Monoidal.hs

```diff
@@ -0,0 +1,193 @@
+  -- no AfterJoin with         A_ReversedLens
+  AfterJoin A_Traversal        A_ReversedPrism    = A_Fold
+  AfterJoin A_Traversal        A_Prism            = A_Traversal
+  -- no AfterJoin with         A_Review
+  AfterJoin A_Traversal        A_Lens             = A_Traversal
+  AfterJoin A_Traversal        A_Getter           = A_Fold
+  AfterJoin A_Traversal        An_AffineTraversal = A_Traversal
+  AfterJoin A_Traversal        An_AffineFold      = A_Fold
+  AfterJoin A_Traversal        A_Fold             = A_Fold
+  AfterJoin A_Traversal        A_Setter           = A_Setter
+
+  -- A_Fold-----
+  AfterJoin A_Fold             An_Iso             = A_Fold
+  -- no AfterJoin with         A_ReversedLens
+  AfterJoin A_Fold             A_ReversedPrism    = A_Fold
+  AfterJoin A_Fold             A_Prism            = A_Fold
+  -- no AfterJoin with         A_Review
+  AfterJoin A_Fold             A_Lens             = A_Fold
+  AfterJoin A_Fold             A_Getter           = A_Fold
+  AfterJoin A_Fold             An_AffineTraversal = A_Fold
+  AfterJoin A_Fold             An_AffineFold      = A_Fold
+  AfterJoin A_Fold             A_Traversal        = A_Fold
+  -- no AfterJoin with         A_Setter
+
+  -- no AfterJoin with A_Setter-----
+  -- A_Traversal-----
+  AfterJoin A_Setter           An_Iso             = A_Setter
+  -- no AfterJoin with         A_ReversedLens
+  -- no AfterJoin with         A_ReversedPrism
+  AfterJoin A_Setter           A_Prism            = A_Setter
+  -- no AfterJoin with         A_Review
+  AfterJoin A_Setter           A_Lens             = A_Setter
+  -- no AfterJoin with         A_Getter
+  AfterJoin A_Setter           An_AffineTraversal = A_Setter
+  -- no AfterJoin with         An_AffineFold
+  -- no AfterJoin with         A_Fold
+
+  -- END GENERATED CONTENT
+
+  -- Every optic kinds can be joined with itself.
+  AfterJoin k k = k
+
+  -- Everything else is a type error.
+  AfterJoin k l = TypeError ('ShowType k
+                       ':<>: 'Text " and "
+                       ':<>: 'ShowType l
+                       ':<>: 'Text " cannot be composed sequentially")
+
+infixr 6 <++>
+class After is m where
+  nothing :: Optic' m is s a
+  (<++>)  :: (Is k m, Is l m, m ~ AfterJoin k l)
+          => Optic' k is s a -> Optic' l is s a
+          -> Optic' m is s a
+
+instance After NoIx A_Traversal where
+  nothing = traversalVL $ \_ s -> pure s
+  t1 <++> t2 = traversalVL $ \f s -> traverseOf t1 f s *> traverseOf t2 f s
+instance After (WithIx ix) A_Traversal where
+  nothing = itraversalVL $ \_ s -> pure s
+  t1 <++> t2 = itraversalVL $ \f s -> itraverseOf t1 f s *> itraverseOf t2 f s
+
+instance After NoIx A_Fold where
+  nothing = foldring (\_ r _ -> r)
+  (<++>) = summing
+instance After (WithIx ix) A_Fold where
+  nothing = ifoldring (\_ r _ -> r)
+  (<++>) = isumming
+
+instance After NoIx A_Setter where
+  nothing = sets (\_ s -> s)
+  s1 <++> s2 = sets (\f s -> over s2 f (over s1 f s))
```

Comment:
There should be a big warning. The resulting `Setter` is lawful only(?) if halves of it don't modify the same parts of underlying structure.

I think that in this case, `<++>` have to be commutative operator. Yet for folds that doesn't make sense.

This structure feels very ad-hoc to me. I'm not sure this is good idea at all to overload this (or `summing` or ...) name.

Reply:
That is exactly the use case I had in mind. Imagine that I want to define several `Traversal`s which works over binary trees. With this API I could define it as:

```haskell
data BinaryTree a = Leaf { _leaf :: a } | Node { _left, _right :: BinaryTree a }

preOrder = leaf <++> left % preOrder <++> right % preOrder
inOrder = left % preOrder <++> leaf <++> right % preOrder
postOrder = left % preOrder <++> right % preOrder <++> leaf
```

Reply:
This is a nice example use case, and could do with appearing in the documentation, whatever API variant we end up with. :smile: